homeassistant/components/sensor/imap.py

```diff
@@ -54,45 +63,107 @@ def __init__(self, name, user, password, server, port):
+        if not self.should_poll:
+            self._idle_loop_task = self.hass.loop.create_task(self.idle_loop())
 
     @property
     def name(self):
         """Return the name of the sensor."""
         return self._name
 
     @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return ICON
+
+    @property
     def state(self):
         """Return the number of unread emails."""
         return self._unread_count
 
-    def update(self):
-        """Check the number of unread emails."""
-        import imaplib
-        try:
-            self.connection.select()
-            self._unread_count = len(self.connection.search(
-                None, 'UnSeen UnDeleted')[1][0].split())
-        except imaplib.IMAP4.error:
-            _LOGGER.info("Connection to %s lost, attempting to reconnect",
-                         self._server)
-            try:
-                self.connection = self._login()
-            except imaplib.IMAP4.error:
-                _LOGGER.error("Failed to reconnect.")
+    @property
+    def available(self):
+        """Return the availability of the device."""
+        return self._connection is not None
 
     @property
-    def icon(self):
-        """Return the icon to use in the frontend."""
-        return ICON
+    def should_poll(self):
+        """Return if polling is needed."""
+        return self._connection and not self._connection.has_capability('IDLE')
+
+    @asyncio.coroutine
+    def connection(self):
+        """Return a connection to the server, establishing it if necessary."""
+        import aioimaplib
+
+        if self._connection is None:
+            try:
+                self._connection = aioimaplib.IMAP4_SSL(
+                    self._server, self._port)
+                yield from self._connection.wait_hello_from_server()
+                yield from self._connection.login(self._user, self._password)
+                yield from self._connection.select()
+            except (aioimaplib.AioImapException, asyncio.TimeoutError):
+                self._connection = None
+
+        return self._connection
+
+    @asyncio.coroutine
+    def idle_loop(self):
+        """Wait for data pushed from server."""
+        import aioimaplib
+
+        while True:
+            try:
+                if (yield from self.connection()):
+                    yield from self.refresh_unread_count()
+                    yield from self.async_update_ha_state()
+
+                    idle = yield from self._connection.idle_start()
+                    yield from self._connection.wait_server_push()
+                    self._connection.idle_done()
+                    yield from asyncio.wait_for(idle, 10)
+                else:
+                    yield from self.async_update_ha_state()
+            except (aioimaplib.AioImapException, asyncio.TimeoutError):
+                self.disconnected()
+
+    @asyncio.coroutine
+    def async_update(self):
```

Comment:
Why do you pull data and have a push?

Reply:
This is because push support (the IDLE capability) is optional for an IMAP server so `should_poll` can be `True` or `False` depending on the server.

I have added a commit making this a bit more readable.